### PR TITLE
Modular testing setup

### PR DESCRIPTION
**Pull request by Su5eD**
## The Issue
With the introduction of modules in modlauncher, existing tests can no longer run due to JUnit running them without BootstrapLauncher. That would usually result in a `ClassCastException` from [ModuleLayerHandler](https://github.com/McModLauncher/modlauncher/blob/9b69c82a176c0896b264ab167813a52adc6f12cb/src/main/java/cpw/mods/modlauncher/ModuleLayerHandler.java#L52), which expects itself to be loaded by a `ModuleClassLoader`.

See also:
- #71

## Launching without BootstrapLauncher
To address the issue, I've modified ModuleLayerHandler to create a new ModuleClassLoader if it doesn't exist already (e.g. when running without BSL). We just pass in `ModuleLayer.boot`, and because none of its modules are referenced by a [`JarModuleReference`](https://github.com/McModLauncher/securejarhandler/blob/41ecce081706f0596abeab8c060fecfb5deb4693/src/main/java/cpw/mods/cl/JarModuleFinder.java#L46), `resolvedRoots` will remain empty and all class loading will be delegated to respective module class loaders (App / Platform).

This is NOT a breaking change, and won't affect dependents.

## Modular Unit tests
For the new test setup, I made the test source set a module and updated the unit tests to follow ML9 usage.
- Cleaned up test code
  - Made test classes and methods package-private
  - Removed unnecessary [ClassTransformer](https://github.com/McModLauncher/modlauncher/pull/101/files#diff-16bb9e6a696484908e9e56707a9177f61bca04861e44eb0e7e90d172c5391b7aL66-L86) 
- `testJarPath` is retrieved from a system property instead of searching the classpath
- Changed target class from `cheese.Puffs` to `cpw.mods.modlauncher.testjar.CheesePuffs` in [TransformingClassLoaderTests](https://github.com/McModLauncher/modlauncher/blob/9c7f00ce2b425a012c8e9d8e5bcfc4a3ad3e4356/src/test/java/cpw/mods/modlauncher/test/TransformingClassLoaderTests.java#L48), as currently we can only generate classes in existing packages
- To prevent module conflicts with launcher tests, testJars has been made a compileOnly dependency, and is resolved at runtime when required

## JMH Benchmark
The [JMH Gradle Plugin](https://github.com/melix/jmh-gradle-plugin) had some issues registering test-only project dependencies, so I replaced it with our own solution, which grants us better control over JMH and result reports. Benchmark classes remain in their own source set, we just create the run task ourselves.
JMH now also runs in modular mode, similarly to tests.

`testJars` has been moved back to a sourceSet as it no longer needs to be a subproject.

### Dependencies
- [x] https://github.com/McModLauncher/securejarhandler/pull/25 Required to launch without BSL

### CODE REVIEW

**Comment by cpw** [comment]
Nice, thank you sir. I'll take a look at merging it in. Having test harnesses back will improve many things, not just ModLauncher.

File: src/main/java/cpw/mods/modlauncher/ModuleLayerHandler.java [hunk 1]
```diff
@@ -49,7 +49,10 @@ SecureJar build() {
     private final EnumMap<Layer, LayerInfo> completedLayers = new EnumMap<>(Layer.class);
 
     ModuleLayerHandler() {
-        completedLayers.put(Layer.BOOT, new LayerInfo(getClass().getModule().getLayer(), (ModuleClassLoader) getClass().getClassLoader()));
+        ClassLoader classLoader = getClass().getClassLoader();
+        ModuleClassLoader cl = classLoader instanceof ModuleClassLoader moduleCl ? moduleCl
+            : new ModuleClassLoader("BOOT", ModuleLayer.boot().configuration(), List.of(ModuleLayer.boot()));
+        completedLayers.put(Layer.BOOT, new LayerInfo(getClass().getModule().getLayer(), cl));
```
Comment (OrionDevelopment) [inline src/main/java/cpw/mods/modlauncher/ModuleLayerHandler.java:55]: Could you please document why this change was made?

**Review by OrionDevelopment — CHANGES_REQUESTED** [review]
Besides a minor comment on the reason for changing the module layer: This looks okey. Could you please add the requested comment?

**Comment by OrionDevelopment** [comment]
Also as you can see from the attached build log, your tests are failing @Su5eD could you take a look and fix those.
[ModLauncher_Pull_Requests_9.1.10.log](https://github.com/McModLauncher/modlauncher/files/8313982/ModLauncher_Pull_Requests_9.1.10.log)

**Comment by Su5eD** [comment]
The TransformingClassLoaderTests class requires [SecureJarHandler#25](https://github.com/McModLauncher/securejarhandler/pull/25) to pass due to the way SecureJarHandler handles package lookups. Please see the PR for the full issue explanation.

**Comment by OrionDevelopment** [comment]
With McModLauncher/Securejarhandler#25 merged what does this need to run @Su5eD ?

**Comment by Su5eD** [comment]
It should run fine now, assuming CI picks up the new SJH release

**Review by OrionDevelopment — APPROVED** [review]

**Pull request merged**